File: cc/053.cc
// Project Euler Problem 053
//
// How many values of C(n,r), for 1 ≤ n ≤ 100, exceed one-million?

#include <cstdio>
#include <vector>

#include "factored.h"

int main(int argc, const char* argv[]) {
  vector<Factored> factorials;
  Factored f(1);
  factorials.push_back(f);
  for (int i = 2; i<=100; ++i) {
    f = f * Factored(i);
    factorials.push_back(f);
  }

  // C(n,r) = n!/(n!*(n-r)!)

  int count = 0;

  for (int n = 1; n <= 100; ++n) {
    Factored num = factorials[n-1];
    if (!(num > 1000000)) {
      continue;
    }
    for (int r = 1; r < n; ++r) {
      if (num / factorials[r-1] / factorials[n-r-1] > 1000000) {
        count++;
      }
    }
  }
  printf("%d\n", count);
}

File: go/problems/053.go
// Project Euler Problem 053
//
// How many values of C(n,r), for 1 ≤ n ≤ 100, exceed one-million?

package problems

import "fmt"

func choicesGt1e6(n int) int {
	choices := 1
	for r := 1; r <= n/2; r++ { // Halfway only: it's a mirror-image
		choices *= (n - r + 1)
		choices /= r
		if choices > 1000000 {
			// All the rest (on both sides of the halfway mark)
			return n + 1 - r*2
		}
	}
	return 0
}

func Problem053() string {
	count := 0
	for n := 23; n <= 100; n++ {
		count += choicesGt1e6(n)
	}
	return fmt.Sprint(count)
}

func init() {
	Functions[53] = Problem053
}
